### PR TITLE
replace url in jad/jar

### PR DESCRIPTION
**Pull request by emord**
https://github.com/dimagi/commcarehq-ansible/pull/336

cc: @calellowitz

### CODE REVIEW

File: corehq/apps/builds/jadjar.py [hunk 1]
```diff
@@ -132,7 +134,7 @@ def pack(self, files, jad_properties=None):
         buffer = StringIO(self.jar)
         with ZipFile(buffer, 'a', ZIP_DEFLATED) as zipper:
             for path in files:
-                zipper.writestr(path, files[path])
+                zipper.writestr(path, files[path].replace(BASE_ADDRESS, J2ME_ADDRESS))
```
Comment (emord) [inline corehq/apps/builds/jadjar.py]: Kind of hacky, but it is nice that we don't have duplicate files running around.

Comment (millerdev) [inline corehq/apps/builds/jadjar.py]: This seems a little dangerous. Is it ever possible that there could be more than one occurrence of `BASE_ADDRESS` within `files[path]`? If so, then this is not safe.

A safer approach (assuming the occurrence of `BASE_ADDRESS` that you're replacing is always at the beginning of the path) would be this:

```py
rel_path = files[path][:len(BASE_ADDRESS)].lstrip("/")  # assume always POSIX paths
new_path = os.path.join(J2ME_ADDRESS, rel_path)
```

Comment (millerdev) [inline corehq/apps/builds/jadjar.py]: update: added assert

Comment (emord) [inline corehq/apps/builds/jadjar.py]: `files[path]` is actually a string (xml) not a path to a file. Although just thought of a better way to do this, so I'll rework it.

File: corehq/apps/builds/jadjar.py [hunk 2]
```diff
@@ -109,6 +112,34 @@ def sign_jar(jad, jar):
...
+            'PostURL',
+            'PostTestURL',
+            'key_server',
+        ]
+        for prop in properties:
+            prop_elem = tree.find("property[@key='" + prop + "']")
+            prop_elem.set('value', transform(prop_elem.get('value')))
+
+
+        for remote in tree.findall("suite/resource/location[@authority='remote']"):
```
Comment (dimagimon) [inline corehq/apps/builds/jadjar.py]: E303 too many blank lines (2)

File: corehq/apps/builds/jadjar.py [hunk 3]
```diff
@@ -1,11 +1,14 @@
...
 import itertools
+import os
+from StringIO import StringIO
+from subprocess import PIPE
+from tempfile import NamedTemporaryFile
 from zipfile import ZipFile, ZIP_DEFLATED
+
 from django.conf import settings
+from lxml import etree
 import shlex
```
Comment (millerdev) [inline corehq/apps/builds/jadjar.py]: nitpick: `shlex` is in the std lib

**Comment by millerdev** [comment]
:+1:

**Comment by nickpell** [comment]
updated to use http.  QAing shortly

File: corehq/apps/builds/jadjar.py [hunk 4]
```diff
@@ -1,14 +1,27 @@
...
+
 from dimagi.utils.subprocess_manager import subprocess_context
 
 
+def _make_address_j2me_safe(address):
+    if settings.J2ME_ADDRESS:
+        return address.replace(
+            settings.BASE_ADDRESS, settings.J2ME_ADDRESS, 1
+        ).replace(
+            "https://%s" % settings.J2ME_ADDRESS, "http://%s" % settings.J2ME_ADDRESS
```
Comment (emord) [inline corehq/apps/builds/jadjar.py]: If this gets deployed to prod before the appropriate redirect code it would force all new builds to use http right?

Comment (nickpell) [inline corehq/apps/builds/jadjar.py]: I'm having ```settings.J2ME_ADDRESS``` default to ```''``` and not replacing when it's an empty string.

**Comment by nickpell** [comment]
hoping to merge today once http://manage.dimagi.com/default.asp?225900 is fixed
@biyeun

File: corehq/apps/app_manager/views/download.py [hunk 5]
```diff
@@ -17,6 +17,7 @@
     AppManagerException, FormNotFoundException
 from corehq.apps.app_manager.util import add_odk_profile_after_build
 from corehq.apps.app_manager.views.utils import back_to_main, get_langs
+from corehq.apps.builds.jadjar import convert_XML_To_J2ME, CONVERTED_PATHS
```
Comment (dimagimon) [inline corehq/apps/app_manager/views/download.py]: F401 'CONVERTED_PATHS' imported but unused

**Comment by nickpell** [comment]
resubmitted to QA, then will be good to merge.

File: corehq/apps/app_manager/views/download.py [hunk 6]
```diff
@@ -231,6 +232,8 @@ def resolve_path(path):
         else:
             _, buffer = obj.get()
             payload = buffer.getvalue()
+        if path in ['profile.xml', 'media_profile.xml']:
```
Comment (nickpell) [inline corehq/apps/app_manager/views/download.py]: @ctsims Can you confirm that ```'profile.xml'``` and  ```'media_profile.xml'``` are only downloaded by j2me?

**Comment by biyeun** [comment]
Looks good. waiting on QA.

File: corehq/apps/app_manager/views/download.py (hunk 6 above)
Comment (nickpell) [inline corehq/apps/app_manager/views/download.py]: (confirmed offline

**Comment by nickpell** [comment]
passed QA

**Comment by nickpell** [comment]
@biyeun gave this the OK, so I'm going to self-merge

**Pull request merged**